File: common/src/main/java/net/eve0415/spigot/WebsocketIntegration/WebsocketEventHandler.java
package net.eve0415.spigot.WebsocketIntegration;

import org.json.JSONException;
import org.json.JSONObject;

import io.socket.client.Manager;
import io.socket.client.Socket;
import io.socket.emitter.Emitter;
import net.eve0415.spigot.WebsocketIntegration.Util.WSIEventState;
import net.eve0415.spigot.WebsocketIntegration.Util.WSIPlatformType;

public class WebsocketEventHandler {
    private final Socket socket;

    public WebsocketEventHandler(final WebsocketManager instance, final Socket socket) {
        this.socket = socket;
        initWebsocketEvents();
    }

    private void initWebsocketEvents() {
        socket.once(Socket.EVENT_CONNECT, new Emitter.Listener() {
            @Override
            public void call(final Object... args) {
                WebsocketManager.getInstance().getWSILogger().info("Connected");
                WebsocketManager.getInstance().isConnected(true);
                loadReconnectEvent();
                configureRoom();
                if (WebsocketManager.getInstance().isStarting()) {
                    try {
                        WebsocketManager.getInstance().send(WSIEventState.STARTING,
                                WebsocketManager.builder().basic().toJSON());
                    } catch (final JSONException e) {
                        WebsocketManager.getInstance().getWSILogger()
                                .error("There was an error trying to send starting status.", e);
                    }
                }
                if (WebsocketManager.getInstance().getPlatformType() == WSIPlatformType.Velocity)
                    WebsocketManager.getInstance().sendServerInfo();
            }
        });

        socket.on(Socket.EVENT_DISCONNECT, new Emitter.Listener() {
            @Override
            public void call(final Object... args) {
                WebsocketManager.getInstance().getWSILogger().info("Disconnected: " + args[0].toString());
                if (args[0].toString().equals("io server disconnect"))
                    socket.connect();
                WebsocketManager.getInstance().isConnected(false);
            }
        });

        socket.on("chat", new Emitter.Listener() {
            @Override
            public void call(final Object... args) {
                try {
                    final JSONObject json = (JSONObject) args[0];
                    WebsocketManager.getInstance().handleWebsocketMessage(json.getString("name"),
                            json.getString("UUID"), json.getString("URL"), json.getString("message"));
                } catch (final JSONException e) {
                    WebsocketManager.getInstance().getWSILogger().error("There was an error trying to parse message",
                            e);
                }
            }
        });
    }

    private void loadReconnectEvent() {
        socket.io().on(Manager.EVENT_RECONNECT, new Emitter.Listener() {
            @Override
            public void call(final Object... args) {
                WebsocketManager.getInstance().getWSILogger().info("Reconnected");
                WebsocketManager.getInstance().isConnected(true);
                configureRoom();
                if (WebsocketManager.getInstance().getPlatformType() == WSIPlatformType.Velocity)
                    WebsocketManager.getInstance().sendServerInfo();
            }
        });
    }

    private void configureRoom() {
        socket.emit("ROOM", WebsocketManager.getInstance().getServerID());
    }
}

File: common/src/main/kotlin/net/eve0415/mc/WebsocketIntegration/WebsocketEventHandler.kt
package net.eve0415.mc.WebsocketIntegration

import io.socket.client.Manager
import io.socket.client.Socket
import net.eve0415.mc.WebsocketIntegration.Enum.WIEventState
import net.eve0415.mc.WebsocketIntegration.Enum.WIPlatformType
import org.json.JSONObject

class WebsocketEventHandler
constructor(private val instance: WebsocketManager, private val socket: Socket) {
  init {
    socket.once(Socket.EVENT_CONNECT) {
      instance.bootstrap.logger.info("Connected")
      instance.isConnected = true
      loadReconnectEvent()
      configureRoom()
      if (instance.isStarting)
        instance.send(WIEventState.STARTING, instance.builder().basic().toJSON())
      if (instance.platformType == WIPlatformType.Velocity) instance.bootstrap.sendServerInfo()
    }

    socket.on(Socket.EVENT_DISCONNECT) {
      fun call(reason: String) {
        instance.bootstrap.logger.info("Disconnected: $reason")
        if (reason == "io server disconnect") socket.connect()
        instance.isConnected = false
      }
    }

    socket.on("chat") {
      val json = it[0] as JSONObject
      instance.bootstrap.handleChatMessage(
        json.getString("name"),
        json.getString("UUID"),
        json.getString("URL"),
        json.getString("message")
      )
    }
  }

  private fun loadReconnectEvent() {
    socket.io().on(Manager.EVENT_RECONNECT) {
      instance.bootstrap.logger.info("Reconnected")
      instance.isConnected = true
      configureRoom()
      if (instance.isStarting)
        instance.send(WIEventState.STARTING, instance.builder().basic().toJSON())
      if (instance.platformType == WIPlatformType.Velocity) instance.bootstrap.sendServerInfo()
    }
  }

  private fun configureRoom() {
    socket.emit("ROOM", instance.serverID)
  }
}
